### PR TITLE
fix: address nil pointer when controller runs with sts and replicas > 1

### PR DESCRIPTION
**Pull request by leoluz**
- fix: address nil pointer when controller runs with sts and replicas > 1
- fix: lint

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

File: util/db/db.go [hunk 1]
```diff
@@ -152,9 +152,8 @@ func StripCRLFCharacter(input string) string {
 func (db *db) GetApplicationControllerReplicas() int {
 	// get the replicas from application controller deployment, if the application controller deployment does not exist, check for environment variable
 	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
-	appControllerDeployment, err := db.kubeclientset.AppsV1().Deployments(db.settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
-
-	if err == nil && appControllerDeployment.Spec.Replicas != nil {
+	appControllerDeployment, _ := db.kubeclientset.AppsV1().Deployments(db.settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
```
Comment (crenshaw-dev) [inline util/db/db.go:155]: Feels a bit dangerous to just discard the error. Should we just selectively ignore 404s and bubble any other errors up?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: util/db/db.go (hunk 1 above)
Comment (leoluz) [reply util/db/db.go]: fair. fixed

**Review by crenshaw-dev — APPROVED** [review]

**Review by ishitasequeira — CHANGES_REQUESTED** [review]

File: util/db/db.go [hunk 2]
```diff
@@ -153,8 +155,14 @@ func (db *db) GetApplicationControllerReplicas() int {
...
 	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
 	appControllerDeployment, err := db.kubeclientset.AppsV1().Deployments(db.settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
-
-	if err == nil && appControllerDeployment.Spec.Replicas != nil {
+	if err != nil {
+		// We should just ignore not found errors as Argo CD controller can
+		// run as a StatefulSet or as a Deployment
+		if !kubeerrors.IsNotFound(err) {
+			log.Warnf("error retrieveing Argo CD controller deployment: %s", err)
+		}
```
[the remark this answers is not in the document]
Comment (leoluz) [reply util/db/db.go]: fixed

File: util/db/db.go [hunk 3]
```diff
@@ -153,8 +155,16 @@ func (db *db) GetApplicationControllerReplicas() int {
 	// get the replicas from application controller deployment, if the application controller deployment does not exist, check for environment variable
 	applicationControllerName := env.StringFromEnv(common.EnvAppControllerName, common.DefaultApplicationControllerName)
 	appControllerDeployment, err := db.kubeclientset.AppsV1().Deployments(db.settingsMgr.GetNamespace()).Get(context.Background(), applicationControllerName, metav1.GetOptions{})
-
-	if err == nil && appControllerDeployment.Spec.Replicas != nil {
+	if err != nil {
+		if kubeerrors.IsNotFound(err) {
+			appControllerDeployment = nil
```
Comment (ishitasequeira) [inline util/db/db.go:160]: We should set `appControllerDeployment = nil` irrespective of whether we find the `appControllerDeployment` or not. If we get an error, we basically did not find the deployment and `appControllerDeployment` will be set as an empty object.

**Review by ishitasequeira — CHANGES_REQUESTED** [review]

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Pull request merged**